File: lib/trends/huya.ts
// Huya — public live category list. No key. Game IDs from Huya's URL scheme (huya.com/g/{id}).
// Falls back to overall top-game list if no category match.

import { wrap, fetchJson, fmtNum } from './_helpers'
import type { TrendQuery, TrendResult, TrendDatapoint } from './types'

const HUYA_CATEGORY_ID: Record<string, number> = {
  'lol': 1, '英雄联盟': 1,
  '王者荣耀': 2336, 'honor of kings': 2336,
  'dota2': 5, 'dota': 5,
  '永劫无间': 6090, 'naraka': 6090,
  '原神': 3203, 'genshin': 3203,
  'apex': 5973, 'apex legends': 5973,
  '绝地求生': 2356, 'pubg': 2356,
  '和平精英': 4438,
  '英雄联盟手游': 6010, 'wild rift': 6010,
  'fps': 0, 'moba': 0,
  '逆水寒': 6620,
  '永劫': 6090,
  '战地': 5887,
  '使命召唤': 5879, 'cod': 5879,
  '我的世界': 660, 'minecraft': 660,
  '梦幻西游': 1366,
}

type LiveListResp = { data?: { datas?: Array<{ nick: string; totalCount: number; gameFullName?: string; introduction?: string }> } }

export async function fetchHuya(query: TrendQuery): Promise<TrendResult> {
  return wrap('huya', query, async () => {
    const category = (query.category ?? '').toLowerCase().trim()
    const gameId = HUYA_CATEGORY_ID[category]
    if (!gameId) throw new Error(`unknown huya category: ${category}`)

    const url = `https://live.huya.com/liveHttpUI/getLiveList?gameId=${gameId}&tagAll=0&page=1`
    const r = await fetchJson<LiveListResp>(url)
    const rooms = r.data?.datas ?? []
    if (rooms.length === 0) throw new Error('empty room list')

    const top = rooms.slice(0, 10)
    const totalViewers = rooms.reduce((s, x) => s + (x.totalCount ?? 0), 0)
    const datapoints: TrendDatapoint[] = top.map(r => ({ label: r.nick, value: r.totalCount, meta: { intro: r.introduction ?? '' } }))

    const summary = `虎牙「${category}」当前 ${rooms.length} 路直播，总热度 ${fmtNum(totalViewers)}（峰值非真人观众数，是虎牙的热度值）。`
    const digest = [
      `# 虎牙直播 · ${category} (gameId ${gameId})`,
      '',
      `**当前直播间**：${rooms.length} 路 · **总热度**：${fmtNum(totalViewers)}`,
      '',
      '## 热度前 10',
      ...top.map((r, i) => `${i + 1}. **${r.nick}** · 热度 ${fmtNum(r.totalCount)} · ${(r.introduction ?? '').slice(0, 40)}`),
    ].join('\n')
    return { summary, digest, datapoints }
  })
}

File: server/stratsquad/trends/huya.py
"""Huya — public live category list."""
from __future__ import annotations

from ..types import TrendQuery, TrendDatapoint, TrendResult
from .base import wrap, fetch_json, fmt_num


HUYA_CATEGORY_ID = {
    "lol": 1, "英雄联盟": 1, "王者荣耀": 2336, "honor of kings": 2336,
    "dota2": 5, "dota": 5, "永劫无间": 6090, "naraka": 6090, "原神": 3203,
    "genshin": 3203, "apex": 5973, "apex legends": 5973, "绝地求生": 2356,
    "pubg": 2356, "和平精英": 4438, "英雄联盟手游": 6010, "wild rift": 6010,
    "逆水寒": 6620, "永劫": 6090, "战地": 5887, "使命召唤": 5879, "cod": 5879,
    "我的世界": 660, "minecraft": 660, "梦幻西游": 1366,
}


async def fetch_huya(query: TrendQuery) -> TrendResult:
    async def body() -> dict:
        category = (query.category or "").lower().strip()
        game_id = HUYA_CATEGORY_ID.get(category)
        if not game_id:
            raise ValueError(f"unknown huya category: {category}")
        url = f"https://live.huya.com/liveHttpUI/getLiveList?gameId={game_id}&tagAll=0&page=1"
        r = await fetch_json(url)
        rooms = ((r.get("data") or {}).get("datas")) or []
        if not rooms:
            raise ValueError("empty room list")
        top = rooms[:10]
        total = sum(x.get("totalCount", 0) for x in rooms)
        datapoints = [TrendDatapoint(label=r["nick"], value=float(r["totalCount"]), meta={"intro": r.get("introduction", "")}) for r in top]
        summary = f"虎牙「{category}」当前 {len(rooms)} 路直播，总热度 {fmt_num(total)}（峰值非真人观众数，是虎牙的热度值）。"
        digest_lines = [
            f"# 虎牙直播 · {category} (gameId {game_id})",
            "",
            f"**当前直播间**：{len(rooms)} 路 · **总热度**：{fmt_num(total)}",
            "",
            "## 热度前 10",
        ]
        for i, x in enumerate(top):
            digest_lines.append(f"{i+1}. **{x['nick']}** · 热度 {fmt_num(x['totalCount'])} · {x.get('introduction', '')[:40]}")
        return {"summary": summary, "digest": "\n".join(digest_lines), "datapoints": datapoints}

    return await wrap("huya", query, body)
